pennylane/gradients/parameter_shift.py

```diff
@@ -327,28 +380,184 @@ def processing_fn(results):
             start = start + num_tapes
 
-            g = _evaluate_gradient(res, data, broadcast, r0, scalar_qfunc_output)
+            g = _evaluate_gradient(tape, res, data, broadcast, r0, scalar_qfunc_output)
 
             grads.append(g)
             # This clause will be hit at least once (because otherwise all gradients would have
             # been zero), providing a representative for a zero gradient to emulate its type/shape.
             zero_rep = qml.math.zeros_like(g)
 
-        for i, g in enumerate(grads):
-            # Fill in zero-valued gradients
-            if g is None:
-                grads[i] = zero_rep
-            # The following is for backwards compatibility; currently, the device stacks multiple
-            # measurement arrays, even if not the same size, resulting in a ragged array.
-            # In the future, we might want to change this so that only tuples of arrays are returned.
-            if getattr(g, "dtype", None) is np.dtype("object") and qml.math.ndim(g) > 0:
-                grads[i] = qml.math.hstack(g)
+        if qml.active_return():
+            for i, g in enumerate(grads):
+                # Fill in zero-valued gradients
+                if g is None:
+                    grads[i] = zero_rep
 
-        return qml.math.T(qml.math.stack(grads))
+            res = tuple(grads)
+        else:
+            res = _post_process_grads(grads, zero_rep)
+
+        return res
 
     return gradient_tapes, processing_fn
 
 
+def _get_var_with_second_order(pdA2, f0, pdA):
+    """Auxiliary function to compute d(var(A))/dp = d<A^2>/dp -2 * <A> *
+    d<A>/dp for the variances"""
+    return pdA2 - 2 * f0 * pdA
+
+
+def _process_pda2_involutory(tape, pdA2, var_idx):
+    """Auxiliary function for post-processing the partial derivative of <A^2>
+    if there are involutory observables.
+
+    If involutory observables are present, ensure they have zero gradient.
+
+    For the pdA2_tapes, we have replaced non-involutory observables with their
+    square (A -> A^2). However, involutory observables have been left as-is
+    (A), and have not been replaced by their square (A^2 = I). As a result,
+    components of the gradient vector will not be correct. We need to replace
+    the gradient value with 0 (the known, correct gradient for involutory
+    variables).
+    """
+    if not qml.active_return():
+
+        m = [tape.observables[i].name not in NONINVOLUTORY_OBS for i in var_idx]
+        m = qml.math.convert_like(m, pdA2)
+        return qml.math.where(qml.math.reshape(m, [-1, 1]), 0, pdA2)
+
+    zero = qml.math.convert_like(0, pdA2)
+    new_pdA2 = []
+    for i, obs in enumerate(tape.observables):
+        if i in var_idx:
+            obs_name = obs.name
+            if isinstance(obs_name, list):
+                obs_involutory = any(n not in NONINVOLUTORY_OBS for n in obs_name)
+            else:
+                obs_involutory = obs_name not in NONINVOLUTORY_OBS

```

Comment:
This part of the code is repeated from `var_param_shift` if I'm not mistaken. You can pass the `non_involutory` variable from `_get_pda2`.

Reply:
Great idea! Updated. :)